File: api/cpponi/cpponi-test/host.cpp
#include <iostream>
#include <string>
#include <errno.h>
#include <vector>
#include <thread>
#include <memory>
#include <sstream>
#include <string>

#include "oelogo.h"
#include "oni.hpp"
#include "onidevices.hpp"

// Dump raw device streams to files?
//#define DUMPFILES

#ifdef DUMPFILES
std::vector<FILE *> dump_files;
#endif

// Windows- and UNIX-specific includes etc
#ifdef _WIN32
#include <windows.h>
//#pragma comment(lib, "liboni")
#include <stdio.h>
#include <stdlib.h>
#else
#include <unistd.h>
#include <pthread.h>
#endif

volatile int quit = 0;
volatile int display = 0;
volatile int display_clock = 0;
int running = 1;

int parse_reg_cmd(const char *cmd, long *values, int len)
{
    char *end;
    int k = 0;
    for (long i = strtol(cmd, &end, 10); cmd != end;
         i = strtol(cmd, &end, 10)) {
        cmd = end;
        if (errno == ERANGE) {
            return -1;
        }

        values[k++] = i;
        if (k == 3)
            break;
    }

    if (k < len)
        return -1;

    return 0;
}

void print_dev_table(oni::device_map_t devices)
{
    // Show device table
    printf(
        "     "
        "+------------------+-------+-------+-------+-------+-------+-----\n");
    printf("     |        \t\t|  \t|Read\t|Reads/\t|Wrt.\t|Wrt./ \t|     \n");
    printf("     |Dev. idx\t\t|ID\t|size\t|frame \t|size\t|frame \t|Desc.\n");
    printf("+----+------------------+-------+-------+-------+-------+-------+--"
           "---\n");

    for (size_t dev_idx = 0; dev_idx < devices.size(); dev_idx++) {

        const char *dev_str = oni_device_str(devices[dev_idx].id);

        printf("|%02zd  |%05zd: 0x%02x.0x%02x\t|%d\t|%u\t|%u\t|%u\t|%u\t|%s\n",
               dev_idx,
               devices[dev_idx].idx,
               (uint8_t)(devices[dev_idx].idx >> 8),
               (uint8_t)devices[dev_idx].idx,
               devices[dev_idx].id,
               devices[dev_idx].read_size,
               devices[dev_idx].num_reads,
               devices[dev_idx].write_size,
               devices[dev_idx].num_writes,
               dev_str);
    }

    printf("+----+-------------------+-------+-------+-------+-------+-------+-"
           "----\n");
}

void data_loop(std::shared_ptr<oni::context_t> ctx)
{
    unsigned long counter = 0;
    int rc = 0;
    const auto dev_map = ctx->device_map();

    try {

        while (rc == 0 && !quit) {

            auto frame = ctx->read_frame();

            //if (display_clock && counter % 100 == 0)
                //std::cout << "\tSample: " << frame.clock() << "\n\n";

            auto data = frame.data<uint16_t>();

#ifdef DUMPFILES 
            fwrite(data.data(), sizeof(uint16_t), data.size(), dump_files[frame.device_index()]);
#endif

            if (display && counter % 1000 == 0) {

                std::cout << "\tDev: " << frame.device_index() << " ("
                          << oni::device_str(dev_map[frame.device_index()].id)
                          << ")\n";

                std::cout << "\tData: [";

                for (const auto &d : data)
                    std::cout << d << " ";

                std::cout << "]\n";
            }

            counter++;
        }

    } catch (const oni::error_t &ex) {
        quit = true;
        std::cerr << ex.what() << "\n";
    }
}

int main(int argc, char *argv[])
{
    try {
        std::cout << oe_logo_med;

        int host_idx = 0;
        std::string driver;

        if (argc != 2 && argc != 3) {
            std::cout << "usage:\n";
            std::cout << "\thost driver [host_index] ...\n";
            exit(1);
        }

        driver = argv[1];

        if (argc == 3)
            host_idx = atoi(argv[2]);

        // Create context
        auto ctx = std::make_shared<oni::context_t>(driver.c_str(), host_idx);

        // Examine device map
        auto dev_map = ctx->device_map();

        // Print device table
        print_dev_table(dev_map);

#ifdef DUMPFILES
        for (int i = 0; i < dev_map.size(); i++) {

            // Open dump files
            std::stringstream ss;
            ss << "idx-" << i << "_id-" << dev_map[i].id << ".raw";
            dump_files.emplace_back();
            dump_files.back() = fopen(ss.str().c_str(), "wb");
        }
#endif

        std::cout << "Max. read frame size: "
                  << ctx->get_opt<uint32_t>(ONI_OPT_MAXREADFRAMESIZE)
                  << " bytes\n";

        oni_size_t block_size = 1024;
        std::cout << "Setting block read size to: " << block_size << " bytes\n";
        ctx->set_opt(ONI_OPT_BLOCKREADSIZE, block_size);

        std::cout << "Block read size: "
                  << ctx->get_opt<oni_size_t>(ONI_OPT_BLOCKREADSIZE)
                  << " bytes\n";

        block_size = 8192;
        std::cout << "Setting write pre-allocation buffer to: " << block_size
                  << " bytes\n";
        ctx->set_opt(ONI_OPT_BLOCKWRITESIZE, block_size);

        std::cout << "Write pre-allocation size: "
                  << ctx->get_opt<oni_size_t>(ONI_OPT_BLOCKWRITESIZE)
                  << " bytes\n";

        std::cout
            << "System clock rate: " << ctx->get_opt<uint32_t>(ONI_OPT_SYSCLKHZ)
            << " Hz\n";

        // Start acquisition
        ctx->set_opt(ONI_OPT_RUNNING, 1);

        // Generate data thread and continue here config/signal handling in
        // parallel
        std::thread tid(data_loop, ctx);

        // Read stdin
        std::string cmd;
        while (cmd != "q") {

            std::cout << "Enter a command and press enter:\n"
                      << "\tc - toggle 1/1000 clock display\n"
                      << "\td - toggle 1/1000 display\n"
                      << "\tt - print device table\n"
                      << "\tp - toggle running register\n"
                      << "\tr - read from device register\n"
                      << "\tw - write to device register\n"
                      << "\tx - issue a hardware reset\n"
                      << "\tq - quit\n"
                      << ">>> ";

            std::getline(std::cin, cmd);

            if (cmd == "p") {
                running = (running == 1) ? 0 : 1;
                ctx->set_opt(ONI_OPT_RUNNING, running);

                if (running)
                    std::cout << "Running.\n";
                else
                    std::cout << "Paused\n";
            } else if (cmd == "c") {
                display_clock = (display_clock == 0) ? 1 : 0;
            } else if (cmd == "d") {
                display = (display == 0) ? 1 : 0;
            } else if (cmd == "r") {
                std::cout << "Read a device register.\n"
                          << "Enter: dev_idx reg_addr\n"
                          << ">>> ";

                // Read the command
                std::string reg_cmd;
                std::getline(std::cin, reg_cmd);

                // Parse the command string
                long values[2];
                auto rc = parse_reg_cmd(reg_cmd.c_str(), values, 2);
                if (rc == -1) {
                    std::cerr << "Error: bad command\n";
                    continue;
                }

                auto dev_idx = (oni_dev_idx_t)values[0];
                auto addr = (oni_size_t)values[1];

                auto val = ctx->read_reg(dev_idx, addr);
                std::cout << "Reg. value: " << val << "\n";
            } else if (cmd == "w") {
                std::cout << "Write to a device register.\n"
                          << "Enter dev_idx reg_addr reg_val\n"
                          << ">>> ";

                // Read the command
                std::string reg_cmd;
                std::getline(std::cin, reg_cmd);

                // Parse the command string
                long values[3];
                auto rc = parse_reg_cmd(reg_cmd.c_str(), values, 3);
                if (rc == -1) {
                    std::cerr << "Error: bad command\n";
                    continue;
                }

                auto dev_idx = (oni_dev_idx_t)values[0];
                auto addr = (oni_reg_addr_t)values[1];
                auto val = (oni_reg_val_t)values[2];

                ctx->write_reg(dev_idx, addr, val);
                std::cout << "Success\n";
            }
        }

        //ctx->write<uint32_t>(8, std::vector<uint32_t>{1, 2});

        // Join data and signal threads
        quit = 1;
        tid.join();

#ifdef DUMPFILES
        // Close dump files
        for (int dev_idx = 0; dev_idx < dev_map.size(); dev_idx++)
            fclose(dump_files[dev_idx]);
#endif
    } catch (oni::error_t err) {
        std::cerr << err.what();
        return -1;
    }

    return 0;
}

File: api/clroni/clroni-test/Host.cs
using System;
using System.Threading;
using System.Runtime.InteropServices;
using oni;

namespace clroni_test
{
    class DataProcessor
    {
        private oni.Context ctx;

        public bool display = false;
        public volatile bool quit = false;
        const int display_downsample = 101;

        public DataProcessor(oni.Context ctx)
        {
            this.ctx = ctx;
        }

        public void CaptureData()
        {
            ulong counter = 0;
           // 
           // int rc = 0;
           // while (rc == 0 && !quit)
           // {
           //     ctx.Write(8, (uint)counter);
           //     Thread.SpinWait(100000); // Sleep(1);
           //     counter++;
           // }

            int rc = 0;
            while (rc == 0 && !quit)
            {
                try
                {
                    var frame = ctx.ReadFrame();

                    if (counter++ % display_downsample == 0)
                    {
                        if (display)
                        {
                            var dat = frame.Data<ushort>();
                            var idx = frame.DeviceIndex();
                            Console.WriteLine("\tDev: {0} ({1})", idx, Device.Name(ctx.DeviceTable[idx].id));
                            Console.WriteLine("\t[{0}]", String.Join(", ", dat));
                        }
                    }

                }
                catch (ONIException ex)
                {
                    Console.Error.WriteLine("Host failed with the following error: "
                                            + ex.ToString());
                    break;
                }
            }
        }
    }

    class Host
    {
        static void Main(string[] args)
        {
            int hw_idx = 0;
            string driver = "";
            if (args.Length == 0 || args.Length > 2) {
                Console.Error.WriteLine("usage:");
                Console.Error.WriteLine("host driver [index]");
                return;
            } else if (args.Length == 1){
                driver = args[0];
            } else if (args.Length == 2) {
                driver = args[0];
                hw_idx = Int32.Parse(args[1]);
            }

            // Get version
            var ver = oni.lib.NativeMethods.LibraryVersion;
            Console.WriteLine("Using liboni version: " + ver);
            bool running = true;

            try
            {

                using(var ctx = new oni.Context(driver, hw_idx))
                {
                    Console.WriteLine("Found the following devices:");
                    foreach (var elem in ctx.DeviceTable.Values) {

                        Console.WriteLine("\t{0}) ID: {1}, Read size: {2}",
                                          elem.idx,
                                          elem.id,
                                          elem.read_size);
                    }

                    // See how big max frames are
                    Console.WriteLine("Max read frame size: "
                                      + ctx.MaxReadFrameSize);

                    // See the hardware clock
                    Console.WriteLine("System clock frequency: "
                                      + ctx.SystemClockHz);

                    // Start acquisition
                    ctx.SetBlockReadSize(4096);
                    ctx.Start();

                    // Start processor in background
                    var processor = new DataProcessor(ctx);
                    var proc_thread
                        = new Thread(new ThreadStart(processor.CaptureData));
                    proc_thread.Start();

                    int c = 's';
                    while (c != 'q') {
                        Console.WriteLine("Enter a command and press enter:");
                        Console.WriteLine("\tc - toggle 1/100 clock display");
                        Console.WriteLine("\td - toggle 1/100 sample display");
                        Console.WriteLine("\tp - toggle stream pause");
                        Console.WriteLine("\tr - enter register command");
                        Console.WriteLine("\tq - quit");
                        Console.Write(">>> ");

                        var cmd = Console.ReadLine();
                        c = cmd[0];

                        if (c == 'p') {
                            running = !running;
                            if (running) {
                                ctx.Start();
                            } else {
                                ctx.Stop();
                                Console.WriteLine("\tPuased.");
                            }
                        } else if (c == 'd') {
                            processor.display = !processor.display;
                        }
                        // else if (c == 'r') {

                        //    printf("Enter dev_idx reg_addr reg_val\n");
                        //    printf(">>> ");

                        //    // Read the command
                        //    char *buf = NULL;
                        //    size_t len = 0;
                        //    rc = getline(&buf, &len, stdin);
                        //    if (rc == -1) { printf("Error: bad command\n");
                        //    continue;}

                        //    // Parse the command string
                        //    long values[3];
                        //    rc = parse_reg_cmd(buf, values);
                        //    if (rc == -1) { printf("Error: bad command\n");
                        //    continue;}
                        //    free(buf);

                        //    size_t dev_idx = (size_t)values[0];
                        //    oni_reg_addr_t addr = (oni_reg_addr_t)values[1];
                        //    oni_reg_val_t val = (oni_reg_val_t)values[2];

                        //    oni_write_reg(ctx, dev_idx, addr, val);
                        //}
                    }

                    // Stop hardware
                    //ctx.Stop();

                    // Join data and signal threads
                    processor.quit = true;
                    proc_thread.Join(200);

                } // ctx.Dispose() is called.

            } catch (ONIException ex) {
                Console.Error.WriteLine("Host failed with the following error: "
                                        + ex.ToString());
                Console.Error.WriteLine("Current errno: "
                                        + Marshal.GetLastWin32Error());
            }
        }
    }
}
